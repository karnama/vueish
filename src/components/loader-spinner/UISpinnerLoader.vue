<template>
    <transition name="progress-spinner" appear>
        <div class="progress-spinner text-brand-400"
             :class="{ 'indeterminate': !determinate }"
             role="progressbar"
             v-bind="$attrs">
            <svg ref="spinnerDraw"
                 class="spinner-draw stroke-current"
                 preserveAspectRatio="xMidYMid"
                 focusable="false"
                 :viewBox="`0 0 ${diameter} ${diameter}`">
                <circle ref="spinnerCircle"
                        class="spinner-circle stroke-current"
                        cx="50%"
                        cy="50%"
                        :r="radius"
                        :stroke-width="stroke + 'px'" />
            </svg>
            <div v-if="$slots.default" class="abs-center">
                <slot />
            </div>
        </div>
    </transition>
</template>

<script lang="ts">
import { computed, defineComponent, ref, watch, onMounted } from 'vue';
import { SetupArg, SetupReturn } from '@/types';

export default defineComponent({
    name: 'SpinnerLoader',

    inheritAttrs: false,

    props: {
        /**
         * The value indicating where the
         * progress is at currently.
         */
        progress: {
            type: Number,
            validator: (value: number) => value >= 0 && value <= 100
        },

        /**
         * Flag indicating that the loader will run
         * until it's set to false.
         */
        determinate: {
            type: Boolean,
            default: false
        },

        /**
         * The diameter of the circle in pixels.
         */
        diameter: {
            type: Number,
            default: 60
        },

        /**
         * The width of the stroke in pixels.
         */
        stroke: {
            type: Number,
            default: 6
        }
    },

    setup(props: SetupArg<{ diameter: number; stroke: number }>): SetupReturn {
        const radius = computed(() => (props.diameter - props.stroke) / 2);
        const circumference = computed(() => Math.PI * 2 * radius.value);
        const strokeDashOffset = computed(() => {
            if (!props.determinate) {
                return String(circumference.value) + 'px';
            }
            if (props.progress) {
                return String(circumference.value * (100 - props.progress) / 100) + 'px';
            }
            return String(circumference.value) + 'px';
        });

        const spinnerDraw = ref<SVGElementTagNameMap['svg']>(null!);
        const spinnerCircle = ref<SVGCircleElement>(null!);

        /**
         * Re-calculate the svg styles.
         */
        const attachSvgStyles = (): void => {
            spinnerDraw.value.style.width = `${props.diameter}px`;
            spinnerDraw.value.style.height = `${props.diameter}px`;
        };

        /**
         * Re-calculate the circle shape styles.
         */
        const attachCircleStyles = (): void => {
            spinnerCircle.value.style.strokeDashoffset = strokeDashOffset.value;
            spinnerCircle.value.style.strokeDasharray = String(circumference.value) + 'px';
            spinnerCircle.value.style.setProperty('--spinner-start-value', String(0.95 * circumference.value));
            spinnerCircle.value.style.setProperty('--spinner-end-value', String(0.2 * circumference.value));
        };

        watch([() => props.diameter, () => props.stroke, () => props.progress], attachCircleStyles);
        watch(() => props.diameter, attachSvgStyles);

        onMounted(() => {
            attachSvgStyles();
            attachCircleStyles();
        });

        return {
            spinnerDraw,
            spinnerCircle,
            radius,
            circumference,
            strokeDashOffset
        };
    }
});
</script>

<style scoped lang="scss">
.progress-spinner {
    position: relative;
    width: max-content;

    &.indeterminate > svg {
        animation: progress-spinner-rotate 2s linear infinite;

        &.spinner-enter,
        &.spinner-leave-to {
            .spinner-draw {
                opacity: 0;
                transform: scale(.1);
            }
        }

        &.spinner-enter-active,
        &.spinner-leave-active {
            transition-duration: .4s;
            animation: none;
        }

        .spinner-circle {
            animation: 4s infinite ease-in-out;
            animation-name: spinner-stroke-rotate;
        }
    }
}

.abs-center {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
}

.spinner-draw {
    transform: scale(1);
    transform-origin: center;
    transition: 200ms ease-in-out;
    will-change: opacity, transform;
}

.spinner-circle {
    fill: none;
    transform-origin: center;
    transition: stroke-dashoffset 200ms ease-in-out;
}

@keyframes progress-spinner-rotate {
    0% {
        transform: rotate(0)
    }
    100% {
        transform: rotate(360deg)
    }
}

@keyframes spinner-stroke-rotate {
    0% {
        stroke-dashoffset: var(--spinner-start-value);
        transform: rotate(0);
    }
    12.5% {
        stroke-dashoffset: var(--spinner-end-value);
        transform: rotate(0);
    }
    12.51% {
        stroke-dashoffset: var(--spinner-end-value);
        transform: rotateX(180deg) rotate(72.5deg);
    }
    25% {
        stroke-dashoffset: var(--spinner-start-value);
        transform: rotateX(180deg) rotate(72.5deg);
    }
    25.1% {
        stroke-dashoffset: var(--spinner-start-value);
        transform: rotate(270deg);
    }
    37.5% {
        stroke-dashoffset: var(--spinner-end-value);
        transform: rotate(270deg);
    }
    37.51% {
        stroke-dashoffset: var(--spinner-end-value);
        transform: rotateX(180deg) rotate(161.5deg);
    }
    50% {
        stroke-dashoffset: var(--spinner-start-value);
        transform: rotateX(180deg) rotate(161.5deg);
    }
    50.01% {
        stroke-dashoffset: var(--spinner-start-value);
        transform: rotate(180deg);
    }
    62.5% {
        stroke-dashoffset: var(--spinner-end-value);
        transform: rotate(180deg);
    }
    62.51% {
        stroke-dashoffset: var(--spinner-end-value);
        transform: rotateX(180deg) rotate(251.5deg);
    }
    75% {
        stroke-dashoffset: var(--spinner-start-value);
        transform: rotateX(180deg) rotate(251.5deg);
    }
    75.01% {
        stroke-dashoffset: var(--spinner-start-value);
        transform: rotate(90deg);
    }
    87.5% {
        stroke-dashoffset: var(--spinner-end-value);
        transform: rotate(90deg);
    }
    87.51% {
        stroke-dashoffset: var(--spinner-end-value);
        transform: rotateX(180deg) rotate(341.5deg);
    }
    100% {
        stroke-dashoffset: var(--spinner-start-value);
        transform: rotateX(180deg) rotate(341.5deg);
    }
}

.progress-spinner-enter-active {
    transition: all 200ms ease-out;
}

.progress-spinner-leave-active {
    transition: all 200ms ease-in;
}

.progress-spinner-enter,
.appear-active-class,
.progress-spinner-leave-to {
    transform: scale(0.1);
    opacity: 0;
}
</style>
