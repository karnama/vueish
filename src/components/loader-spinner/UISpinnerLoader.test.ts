import UISpinnerLoader from './UISpinnerLoader.vue';
import { mount } from '@vue/test-utils';

const props = {
    progress: 38,
    steps: 76,
    stroke: 10,
    diameter: 75,
    determinate: true
};

describe('UISpinner', () => {
    it('should render', () => {
        const wrapper = mount(UISpinnerLoader);

        expect(wrapper).toBeDefined();
        expect(wrapper.exists()).toBe(true);
    });

    it('should be display the given slot', () => {
        const wrapper = mount(UISpinnerLoader, {
            slots: {
                default: '<div id="default-slot-content" />'
            }
        });

        expect(wrapper.find('#default-slot-content').exists()).toBe(true);
    });

    it('should display with the right dimensions', () => {
        const wrapper = mount(UISpinnerLoader, { props });

        expect(wrapper.find('svg').attributes()['style']).toBe('width: 75px; height: 75px;');
    });

    it('should calculate the radius of the circle correctly', () => {
        const wrapper = mount(UISpinnerLoader, { props });

        const radius = String((props.diameter - props.stroke) / 2);
        expect(wrapper.find('circle').attributes()['r']).toBe(radius);
    });

    it('should dynamically update the stroke width', async () => {
        const wrapper = mount(UISpinnerLoader, { props });

        expect(wrapper.find('circle').attributes()['stroke-width']).toBe(String(props.stroke) + 'px');
        await wrapper.setProps({ stroke: 5 });
        expect(wrapper.find('circle').attributes()['stroke-width']).toBe('5px');
    });

    it('should dynamically update the progress', async () => {
        const getStrokeOffsetStyle = (progress?: number) => {
            progress = progress ?? props.progress;
            const radius = (props.diameter - props.stroke) / 2;
            const circumference = Math.PI * 2 * radius;
            const step = circumference / Number(props.steps);

            return `stroke-dashoffset: ${circumference - step * progress}px`;
        };

        const wrapper = mount(UISpinnerLoader, { props });
        const circle = wrapper.find('circle');

        expect(circle.attributes()['style']).toContain(getStrokeOffsetStyle());
        await wrapper.setProps({ progress: 99 });
        expect(circle.attributes()['style']).toContain(getStrokeOffsetStyle(99));
    });

    it('should dynamically update the diameter', async () => {
        const radius = (diameter: number) => String((diameter - props.stroke) / 2);
        const wrapper = mount(UISpinnerLoader, { props });

        expect(wrapper.find('circle').attributes()['r']).toBe(radius(props.diameter));
        await wrapper.setProps({ diameter: 50 });
        expect(wrapper.find('circle').attributes()['r']).toBe(radius(50));

        expect(wrapper.find('svg').attributes()['viewBox']).toBe('0 0 50 50');
    });

    it('should not apply default colour classes when inherit-color is passed', async () => {
        const wrapper = mount(UISpinnerLoader);

        expect(wrapper.find('.ui-spinner-loader').classes('text-brand-400')).toBe(true);
        await wrapper.setProps({ inheritColor: true });
        expect(wrapper.find('.ui-spinner-loader').classes('text-brand-400')).toBe(false);
    });
});
