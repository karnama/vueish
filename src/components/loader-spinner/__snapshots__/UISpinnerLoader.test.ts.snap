// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UISpinner should render correctly 1`] = `
<transition-stub>
  <div
    class="ui-spinner-loader text-brand-400"
    role="progressbar"
  >
    <svg
      class="spinner-draw stroke-current"
      focusable="false"
      preserveAspectRatio="xMidYMid"
      style="width: 75px; height: 75px;"
      viewBox="0 0 75 75"
    >
      <circle
        class="spinner-circle stroke-current"
        cx="50%"
        cy="50%"
        r="32.5"
        stroke-width="10px"
        style="stroke-dashoffset: 102.10176124166827px; stroke-dasharray: 204.20352248333654px; --spinner-start-value: 193.9933463591697; --spinner-end-value: 40.840704496667314;"
      />
    </svg>
    <!--v-if-->
  </div>
</transition-stub>
`;

exports[`UISpinner should render correctly when set to determinate 1`] = `
<transition-stub>
  <div
    class="ui-spinner-loader indeterminate text-brand-400"
    role="progressbar"
  >
    <svg
      class="spinner-draw stroke-current"
      focusable="false"
      preserveAspectRatio="xMidYMid"
      style="width: 75px; height: 75px;"
      viewBox="0 0 75 75"
    >
      <circle
        class="spinner-circle stroke-current"
        cx="50%"
        cy="50%"
        r="32.5"
        stroke-width="10px"
        style="stroke-dashoffset: 204.20352248333654px; stroke-dasharray: 204.20352248333654px; --spinner-start-value: 193.9933463591697; --spinner-end-value: 40.840704496667314;"
      />
    </svg>
    <!--v-if-->
  </div>
</transition-stub>
`;
